waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts

```diff
@@ -28,38 +30,48 @@ export function getSignupRoute({
         req: Request<{ email: string; password: string; }>,
         res: Response,
     ): Promise<Response<{ success: true } | { success: false; message: string }>> {
-        const userFields = req.body;
-        ensureValidArgs(userFields);
+        const fields = req.body;
+        ensureValidArgs(fields);
         
-        userFields.email = userFields.email.toLowerCase();
-
-        const existingUser  = await findUserBy({ email: userFields.email });
-        // User already exists and is verified - don't leak information
-        if (existingUser && existingUser.isEmailVerified) {
-            await doFakeWork();
-            return res.json({ success: true });
-        } else if (existingUser && !existingUser.isEmailVerified) {
-            if (!isEmailResendAllowed(existingUser, 'emailVerificationSentAt')) {
-                return res.status(400).json({ success: false, message: "Please wait a minute before trying again." });
+        const existingAuthIdentity = await findAuthIdentity("email", fields.email);
+        if (existingAuthIdentity) {
+            const providerData = deserializeProviderData<'email'>(existingAuthIdentity.providerData);
+            // User already exists and is verified - don't leak information
+            if (providerData.isEmailVerified) {
+                await doFakeWork();
+                return res.json({ success: true });
+            } else if (!providerData.isEmailVerified) {
+                if (!isEmailResendAllowed(providerData, 'emailVerificationSentAt')) {
+                    return res.status(400).json({ success: false, message: "Please wait a minute before trying again." });
+                }
+                // User exists but is not verified - delete the user and create a new one
+                await deleteUserByAuthId(existingAuthIdentity.authId);
```

Comment:
I'll create a flow chart for the email signup process, that feels like the best way to document it 👍

Reply:
<img width="1669" alt="shapes at 23-12-15 17 38 09" src="https://github.com/wasp-lang/wasp/assets/2223680/aa32437c-ead9-4664-b175-f08730c4b047">

Reply:
Thanks, this is nice, we should have in the docs maybe!

So this is indeed how I understood it, so my questions remain open:
1. What if developer has allowed unverified email? Anybody can come and delete my account by trying to sign up as me. That is a problem. We can't have allowed unverified email and also allow a signup attempt to delete existing account with unverified email.
Simplest solution that comes to my mind is not allowing unverified email state, except for maybe during development.
2. What about that attack where someone signs up right after me and I verify their account by accident, that I described above in this comment? There is 1 minute period to help with that, but still, I can imagine somebody pulling it off right after 1 minute.

Also, I said that we should document intention -> graph is nice, we could use this, but we should also document intention in the code,  in places where ti is not obvious. Here where I commented is not obvious for sure, and maybe some other places. You do have some comments but they're not rich enough.

Reply:
I've realised that we didn't properly handle the unverified email signup! Thanks for thinking this through.

**Before:** we wouldn't even take into account if the `allowVerifiedEmailLogin` setting was turned on or off. This was completely wrong and the scenario of somebody deleting your account was happening.

**Now**: if there is an existing account with an unverified email, and we allow unverified login -> throw an error `"Hey, you can't signup with an existing email"`

**About the attack where somebody signs up at the same as you with your email** -> that's something that can happen. It's very unlikely, but it's possible. One thing we can do is use additional verification tokens we keep in the DB e.g. `123` which get sent to your email. If you confirm the email with `123`, great, it was you who made that account. If somebody tried to sneak past you, another token would be generated e.g. `abc` and when you click on the link with the old token, the verification wouldn't happen. This reduces the security risk to annoyance which is okay IMHO?

Reply:
I've added the `emailVerificationToken` and `passwordResetToken` fields in the DB. This should help with the "somebody signed up with your email and you confirmed it". But you can still accidentally click on the new link 🤷 

https://github.com/wasp-lang/wasp/pull/1583/commits/5daf665ecf64eef130add318323e99e6a5407817#diff-1bfc66a799b7555ee9cb68328bc71f702449dd83a80b870af6b784139089a0eaR32

Reply:
Let's call this attack "planted email verification" (PEV).

What about period we give them to verify their email address -> if it is too short, it becomes an annoyance, but if it is long, it allows them to occupy an email address that might not be theirs for a long time. What about PEV? Does short or long period help? Short period probable makes PEV a bit easier because you are more likely to have expired link and therefore be tempted to click that new malicious link. Hm.
But this is not really very important, it is just small considerations.

But what about this idea: one thing that is unique to them is their password. What if we ask them to again input their password once they click on reset link? So it is not enough to just click on reset link -> instead, they then need to provide the password once again, and only then we verify their email! If we do this, we solved it. Because the attacker will have another password, and then the email owner will won't be able to confirm that reset link. Yeah, I think we should 100% do this.

What I wanted to suggest is that you research a bit more about the best practices here, what is the recommended way to do email verification flow, and while I would still recommend we do that (what does SO or OWASP say?), I am not pretty confident what I wrote above is the solution.

Reply:
Oh and let's document this well in the code, as we discussed, to capture not only what we do, but why we do it that way. We can have either a special .md explaining how we do it, or just quite rich comments in the code (I think I prefer the later, unless you want to have diagrams, then the former (.md) makes more sense).

Reply:
I've researched what others do regarding this type of attack and it seems that they don't worry themselves about it that much.

- https://cheatsheetseries.owasp.org/cheatsheets/Forgot_Password_Cheat_Sheet.html
- https://auth0.com/docs/manage-users/user-accounts/verify-emails
- I've created an account on Facebook twice in a new and it just sent me extra emails

I think this will be annoying for the users and it's not something that happens commonly in the wild:
>  So it is not enough to just click on reset link -> instead, they then need to provide the password once again

But let's look at our flow one more time:
1. User signs up
2. They receive an email verification email
3. They click a link and verify the email
4. They have to **log in** with their password to access the account

In a way they are "using the password" to confirm their email. Because ...  if the password doesn't work, they'll simply reset the password and have access to their account once again. 

I feel like we are optimising for a rare attack and should reduce the scope of this PR not to include this extra work so we can close it out and move on.

Reply:
I'll write the extra Markdown file with the email auth explained 👍

Reply:
You said you researched what others do, but I didn't get what they do? Could you summarize it for me since you already read these links?
Facebook sends extra emails -> hm, so they don't even wait for the grace period? How is that? So your older account just gets deleted immediately? Do they allow you to do anything with unverified email?

I agree we shouldn't overengineer it, but I don't think we can make the decision what is overengineering or what is not until we understand this problem fully. I think we are much closer to it now, but I would like to understand the conclusions of your research first.

Since we are a web framework implementing Auth for others, that follows best practices, I think it is vital that we do it very well, or we can break users' trust into us. We are going to review all our security practices once again when we come to 1.0, but since we are not in this and have all the context, I think this is a good moment to do as much as we can right now. And it is not like we are doing much here, we are merely learning thoroughly what are the best practices for verification links and exploring possible solutions.

You are right that they have to log in later anyway, so yeah, at that point they will realize that something is wrong and can reset the password. But what if they don't log in immediately, allowing that other person to in the meantime use their email and therefore represent themselves as them? I still don't like that idea. And sounds to me like we can solve it simply by asking for password once again, which doesn't sound like much of a hassle to me, they anyway set up that password just recently. We can even log them in immediately based on that, so they don't have to log in after that, in that case it is not added hassle at all.

I would like to hear conclusions of your research, then we can decide! We can do nothing (not big fan of that at the moment but that might change), we can do something (currently I vote for this but let's see) or we can open issue to do something a bit later (I could live with this).

Reply:
> You said you researched what others do, but I didn't get what they do?

I did mention what was the point of the research: "it seems that they don't worry themselves about it that much"

> Could you summarize it for me since you already read these links?

I don't see the point in me summarising the links they serve as extra details if you are curious. I did write a small doc about Auth0, Supabase and NextAuth if you are interested: https://www.notion.so/wasp-lang/Email-verification-flow-c2958201f6df4a15a1bdc0e2b0a8a9f5?pvs=4

> Facebook sends extra emails -> hm, so they don't even wait for the grace period? 

I haven't seen the grace period no.

> How is that?

I can't speak for Facebook, but I guess they didn't concern with that themselves.

> So your older account just gets deleted immediately?

I'm not sure what happens in the background, but you get another email.

> Do they allow you to do anything with unverified email?

You are required to verify the email before going forward.

> And it is not like we are doing much here, we are merely learning thoroughly what are the best practices for verifying email and are exploring possible solutions.

We are expanding the scope of the initial PR a bit too much IMHO, I'd rather park this discussion and create an issue.

> You are right that they have to log in later anyway, so yeah, at that point they will realize that something is wrong and can reset the password. But what if they don't log in immediately, allowing that other person to in the meantime use their email and therefore represent themselves as them? I still don't like that idea. And sounds to me like we can solve it simply by asking for password once again, which doesn't sound like much of a hassle to me, they anyway set up that password just recently. We can even log them in immediately based on that, so they don't have to log in after that, in that case it is not added hassle at all.

I feel like we are inventing something new here that doesn't need inventing. Especially inside of this PR, as a separate issue with it's own RFC and a proper offline discussion, that would make more sense.

> Ask them to sign up with just an email, no other info. Don't ask for password yet.

This would require changes to the Auth UI and would require additional page to be defined for the signup flow which is really expanding the scope of this PR which I don't feel like is the right thing to do. Let's create an issue that says "Preventing signup attacks" or something similar.

Reply:
https://github.com/wasp-lang/wasp/issues/1614

Reply:
> > You said you researched what others do, but I didn't get what they do?
> 
> I did mention what was the point of the research: "it seems that they don't worry themselves about it that much"

Ah, well that does give me some info, but not much to be honest. "it seems" and "that much" are quite inconcrete. I would love to have a more detailed analysis of what others do, and why.

> 
> > Could you summarize it for me since you already read these links?
> 
> I don't see the point in me summarising the links they serve as extra details if you are curious. I did write a small doc about Auth0, Supabase and NextAuth if you are interested: https://www.notion.so/wasp-lang/Email-verification-flow-c2958201f6df4a15a1bdc0e2b0a8a9f5?pvs=4

I meant just give me some more details so I don't have to do the research on my own, since you didn't give me much info really. What does it mean thy don't care that much, what do they do?

> 
> > Facebook sends extra emails -> hm, so they don't even wait for the grace period?
> 
> I haven't seen the grace period no.
> 
> > How is that?
> 
> I can't speak for Facebook, but I guess they didn't concern with that themselves.
> 
> > So your older account just gets deleted immediately?
> 
> I'm not sure what happens in the background, but you get another email.
> 
> > Do they allow you to do anything with unverified email?
> 
> You are required to verify the email before going forward.

Ok, so the fact that you are required to verify the email before going forward is crucial here, because it changes things quite a bit. This important information. I also wonder if they ask you for password before you verify the email, or later? That also affects it.

I asked a lot of questions here now, but I would if an analysis like this was done by you, so I don't have to pull out the information from you.

> 
> > And it is not like we are doing much here, we are merely learning thoroughly what are the best practices for verifying email and are exploring possible solutions.
> 
> We are expanding the scope of the initial PR a bit too much IMHO, I'd rather park this discussion and create an issue.

Why is that, why do you think this doesn't fit into scope? Since we want to ship this new solution very soon, I also want it to move as fast as possible, but on the other hand, I don't want us to ignore best practices. If you wouldn't do this now, when would you do it? As I mentioned above, I might be ok with postponing this, but I first want to understand the situation we are in.

> 
> > You are right that they have to log in later anyway, so yeah, at that point they will realize that something is wrong and can reset the password. But what if they don't log in immediately, allowing that other person to in the meantime use their email and therefore represent themselves as them? I still don't like that idea. And sounds to me like we can solve it simply by asking for password once again, which doesn't sound like much of a hassle to me, they anyway set up that password just recently. We can even log them in immediately based on that, so they don't have to log in after that, in that case it is not added hassle at all.
> 
> I feel like we are inventing something new here that doesn't need inventing. Especially inside of this PR, as a separate issue with it's own RFC and a proper offline discussion, that would make more sense.

Yeah, maybe that makes sense. But we wouldn't get here if I wasn't pushing all these questions: the problem is that I can't judge when to stop asking them, unless I do the thorough analysis myself, which I want to avoid as it would take a lot of time vs reviewing. So instead, I am asking questions, and expecting you to up your analysis game accordingly, so we can then decide where to draw the line.

> 
> > Ask them to sign up with just an email, no other info. Don't ask for password yet.
> 
> This would require changes to the Auth UI and would require additional page to be defined for the signup flow which is really expanding the scope of this PR which I don't feel like is the right thing to do. Let's create an issue that says "Preventing signup attacks" or something similar.

I don't mind doing all this later if it makes sense, but what is the plan for it, are we ok to release without it? Feels to me like we still want to have those improvements for 12.0 since we are already doing so many changes, right?

Reply:
We agreed offline to handle this and the overall signup flow in a subsequent PR we'll do shortly after this one 👍